### ISSUE TITLE
How would YOU bypass Haven?

### ISSUE THREAD

**Issue by fat-tire** [guardianproject/haven#171]
As a companion to #62 ("_What, if anything, can we do, to combat evil uses of Haven?_"), I think there may be some value in considering/brainstorming practical attacks on Haven-equipped devices.  Doing so may lead to refinements in the program's operation or design to improve its effectiveness.

So you're an evil maid.   It's 3 am.  While your target is out all night drinking, their laptop is in back in the hotel room on a desk, closed.  A plugged-in phone, older model, is resting there on the shut lid.  The phone is attached to the wall via a microUSB cable.  You strongly suspect it's running Haven, and that it's running with all sensors and features enabled.

You do not have the phone # of the Haven device or any info about its SIM card (if it has one) or software running (including Android version, though you can assume latest stock) or any accounts that may be logged in.  But you DO have access to the hotel wifi network, which uses WPA2.  For fun, let's say the connection is vulnerable to [KRACK](https://www.krackattacks.com/).

You want to image the laptop and modify the hardware before the owner returns in 5 hours.  The goal is to NOT warn the owner that anything untoward happened.  You do not want them to receive any texts, SMS, or other notifications during the attack.  Nor do you want to leave any physical or digital evidence, including pictures, sounds, warnings, logs, notifications, or scorch marks on the Haven device when the target returns.  They should have no clue you were ever there.

![Hmmm](https://media.giphy.com/media/9J8gnvAxmDFbG/giphy.gif)

_What do you do?_

I don't have any particular rules except let's say you're unable to completely take out the local wifi for the whole hotel (...but maybe you can jam it for just a small area-- it might be interesting to think about how Haven might behave then... should it log connectivity dropouts or failed connections for example?  Should it expect some kind of unspoofable confirmation that the notification SMS/SIgnal message has been received?)

Is there a social engineering type of attack that might lead a person to think "Oh, Haven crashed" or "the battery ran out. I must have forgotten to plug it in" or "this wasn't an attack, just the AC turning on" to make them think this was just a normal screw up?

If you thought of an attack, what might Haven developers or the device owner need to do to counter it? Is there a software solution?  Is there a placement of the Haven device that might make it less (or more) vulnerable?

I'll start with some dumb ideas:  1.  `adb connect` to the device IP to see if it was mistakenly left on to shut it off remotely (does Haven log & notify the owner on startup and shutdown?).  2. Maybe try a [wifi exploit](https://arstechnica.com/information-technology/2017/07/broadcom-chip-bug-opened-1-billion-phones-to-a-wi-fi-hopping-worm-attack/).   3.  Magic EMP gun to reset only the phone while leaving the laptop intact and untouched.

Your turn.  Stupid ideas are encouraged.  File [new issues](https://github.com/guardianproject/haven/issues/new) as necessary.

**Comment by E3V3A** [comment]
How to bypass the bypasser passerby?

Before anything, you'd run an IMSI-catcher to get the phone number to device. Then run `nmap` on the local WiFi to search for Android and possibly correlate findings with API version. If the API seem outdated and not-patched, then try to run some attacks, including malware/blocking SMS. If that is not working, you'd put the entire table in a Faraday-cage, disconnect the USB charger and replace with a powered USB hub. Then `adb connect` and see if anything else is accessible. If all fails, just brake the phone with the EMP gun inside the Faraday-cage, as the user would never know what happened to it. (Just keep the gun away from the laptop.) 

Unless the attack was 100% successful (full ownership with shutdown/modification/replacement of Haven or sensors), this attack would detectable, which is why this project is cool.

But then again, if someone would need the info in that PC so bad, they'd just take it! So yeah, don't ever leave your PC in your hotel room, if you have any sensitive info on it.

**Comment by leonardoporpora** [comment]
1. Attack: jamming of the communication. Counterattack:  in case of a
jam attack it would be useful to have a heartbeat (as Micah Lee said) generated by the
phone. If the acknowledgement of the heartbeat is not received by the
phone, it will start to log the events until correct reception of the
messages to the server is recovered. Example of failed acknowlegement
may be when connection is missing and Signal alerts that message
delivery was not possible.

2. Attack: phone replacement. Counterattack: Haven should link the user
telephone number signed into Haven phone to the phone MAC address. This
check should be repeeated at a rate of x minutes because if Haven
creates a file with the check of the MAC address and then the phone is
cloned there would be no alert from Haven, whereas if the hash is
checked and it doesn't correspond to the one hashed in the file the user
should be alerted.

3. Additional protection: Haven should create an encrypted volume within
the photos, recordings and logs because if someone jam the Wi-Fi and
connect the phone to a PC than he could modify/delete all the logs and
files. The best tool to do this for me is http://sovworks.com/eds/

4. Additional protection: there must be a way to disable an eventual
remote shutdown of the phone. And I also think that must be useful that
the heartbeat doesn't only say if Haven is active or not that is very
very important to know but also to say the percentage of the battery so
that the user is aware that his phone in short is going to switch off.

**Comment by leonardoporpora** [comment]
Let me know what you think guys about what I wrote @n8fr8 @fat-tire @mig5 @E3V3A

**Comment by n8fr8** [comment]
I think the heartbeat #54 feature is critical to resisting most attacks, and should be prioritized (hint, hint!).

I also agree, we should implement encrypted storage, since the theoretical attack that rewrites the log (essentially) by modifying the media on disk to show something non-malicious ("oh it was just a cat!"), is plausible. Guardian Project already has a well supported, easy to implement library called IOCipher that does this: https://guardianproject.info/code/iocipher/ and would not take much to implement.

**Comment by leonardoporpora** [comment]
I think is also much important to link the user phone number signed into Haven to the MAC address. Take an hash to be sure that the phone replacement didn’t take place. What do you think? @n8fr8

**Comment by n8fr8** [comment]
Regarding the heartbeat, there is the question of the length of the interval between pings. What can an attacker do in 1 minute or 5 minutes?

Also, back to @leonardoporpora comment about swapping the phone... we do need to implement the security code / key verification feature in Signal so that at least you can be assured it is the same instance of Signal, and not a new device swapped in via an SS7 attack to take over the account. Adding in some method to verify the device MAC address, or a PGP private key signature might also be useful, but more work.

**Comment by leonardoporpora** [comment]
Yes I agree that it would be useful maybe with PGP private key but it requires much more work than signal key verification code. Maybe we can do it in future

**Comment by leonardoporpora** [comment]
@n8fr8 anyway rethinking about MAC address verification I think it should be implemented because if people won’t use Signal service but Email service or simply sms they would not be alerted. I’m aware that this feature requires time to be implemented but I think that it’s an important one.

**Comment by leonardoporpora** [comment]
Maybe we could implement with the heartbeat. For example every x minutes when it sends the heartbeat it could says that it’s still active and that the MAC address corrispondes to the right one. I think that both the heartbeat and MAC address verification are important because a user of Haven is not obliged to use signal so signal verification is not enough.
